proposals/NNNN-constructors.md

```diff
@@ -0,0 +1,126 @@
+# Constructors
+
+* Proposal: [NNNN](NNNN-constructors.md)
+* Author(s): [Chris Bieneman](https://github.com/llvm-beanz)
+* Sponsor: [Chris Bieneman](https://github.com/llvm-beanz)
+* Status: **Under Consideration**
+* Planned Version: 202y
+
+## Introduction
+
+As user defined data structures get more common and complex, the absence of
+constructors and destructors is more and more awkward. This proposal seeks to
+bring the full features of C++ initialization and destruction into HLSL.
+
+## Motivation
+
+Constructors are incredibly useful for ensuring data initialization and
+implementing common patterns like Resource-Acquisition-Is-Initialization (RAII).
+
+## Proposed solution
+
+Introduce the following features to HLSL:
+* C++ Constructors and Destructors
+* C++ 11 Delegating Constructors
+* C++ 11 Default and Deleted methods
+* C++ 11 Default initializers
+* Clang __attribute__((constructor)) and __attribute__((destructor))
+
+Constructor syntax will match C++ 11 constructor syntax including support for
+delegating constructors. It will be available on all user-defined data types
+that support constructors in C++.
+
+## Detailed design
+
+### Global Construction
+
+HLSL global constructors will be ordered following C++ conventions.
+Constant-evaluated initializers occur first, followed by non-constant
+initializers which are ordered by declaration order in the translation unit.
+
+Each entry function will call the global constructors before executing any
+further instructions and the global destructor immediately before returning.
+
+In library shaders, the entry translation unit's global constructors are
+executed after the linked library global constructors, and the entry's global
+destructors are executed before linked library global destructors.
+
+Non-trivial global constructors **will not** be stripped from library shaders.
+Meaning, an unused global with a non-trivial constructor will be linked into the
+final linked shader binary. This conforms to the C++ standard definitions for
+static storage duration (see: ISO C++ Standard _basic.stc.static_).
+
+Due to the inherent performance drawbacks of global construction, warnings will
+be issued by default for any code that introduces a non-trivial global
+constructor.
+
+### Constructing Groupshared
+
+Constructor initialization of `groupshared` objects will be followed by a memory
+barrier and group sync. This comes with inherent performance costs, so shader
+authors should use constructors with `groupshared` objects sparingly.
+
+A `groupshared` user defined object that does not have a constructor, or has a
+constructor that does not initialize any of its fields, will not incur the
+memory barrier.
+
+Due to the inherent performance drawbacks of `groupshared` construction,
+warnings will be issued by default for any code that introduces a `groupshared`

```

Comment:
This block makes it seem like we don't want to support groupshared constructors. If thats the case why a warning and not an error? 

If I'm off base maybe we should enumerate supported vs unsupported group shared cases?

Reply:
I think we want to support them. They can be super useful. We just want to also make sure users understand the cost of them.